### PR TITLE
Fix kaiser_window for lower precision data types on CPU

### LINKED ISSUE #117230: ```torch.kaiser_window```: Internal assert failed
### 🐛 Describe the bug

Test code:
```python
import torch

torch.kaiser_window(2, device=torch.device('cpu'), dtype=torch.bfloat16, layout=torch.strided, requires_grad=False)
```
   
Error log:
```
Traceback (most recent call last):
  File "/home/yonghyeon/pytorch_err_case_py/kaiser_window.py", line 3, in <module>
    torch.kaiser_window(2, device=torch.device('cpu'), dtype=torch.bfloat16, layout=torch.strided, requires_grad=False)
RuntimeError: !needs_dynamic_casting<func_t>::check(iter) INTERNAL ASSERT FAILED at "/home/yonghyeon/pytorch/pytorch-asan/aten/src/ATen/native/cpu/Loops.h":310, please report a bug to PyTorch.
```
   
Test code (c++):
```cpp
#include <torch/torch.h>

int main() {
    c10::optional<c10::Device> device = c10::optional<c10::Device>(c10::DeviceType::CPU);

    c10::optional<c10::ScalarType> dtype = c10::optional<c10::ScalarType>(torch::kBFloat16);

    c10::optional<c10::Layout> layout = c10::optional<c10::Layout>(c10::Layout::Strided);

    c10::optional<bool> requires_grad = c10::optional<bool>(false);

    c10::optional<bool> pinned_memory = c10::nullopt;

    c10::optional<c10::MemoryFormat> memory_format = c10::optional<c10::MemoryFormat>(c10::MemoryFormat::ChannelsLast);

    c10::optional<c10::MemoryFormat> merge_memory_format = c10::nullopt;

    auto options = 
        c10::TensorOptions()
        .device(device)
        .dtype(dtype)
        .layout(layout)
        .requires_grad(requires_grad)
        .pinned_memory(pinned_memory)
        .memory_format(memory_format)
        .merge_memory_format(merge_memory_format);

    at::kaiser_window(2, options);

    return 0;
}
```

Error log:
```
terminate called after throwing an instance of 'c10::Error'
  what():  !needs_dynamic_casting<func_t>::check(iter) INTERNAL ASSERT FAILED at "/home/yonghyeon/pytorch/pytorch-asan/aten/src/ATen/native/cpu/Loops.h":310, please report a bug to PyTorch. 
Exception raised from cpu_kernel at /home/yonghyeon/pytorch/pytorch-asan/aten/src/ATen/native/cpu/Loops.h:310 (most recent call first):
frame #0: <unknown function> + 0x1e7a6e (0x7f28ab910a6e in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #1: <unknown function> + 0x1e7a20 (0x7f28ab910a20 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #2: <unknown function> + 0x1e79b0 (0x7f28ab9109b0 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #3: <unknown function> + 0x1e7860 (0x7f28ab910860 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #4: <unknown function> + 0x1e9286 (0x7f28ab912286 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #5: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x1b7 (0x7f28ab90e4d7 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #6: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x259 (0x7f28ab907ca9 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #7: <unknown function> + 0x2cb303e (0x7f28ae7c203e in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #8: <unknown function> + 0x19a073d5 (0x7f28c55163d5 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #9: <unknown function> + 0x19a02ba6 (0x7f28c5511ba6 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #10: <unknown function> + 0x19a01f7b (0x7f28c5510f7b in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x19a0168b (0x7f28c551068b in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #12: <unknown function> + 0x503aa13 (0x7f28b0b49a13 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #13: at::native::kaiser_window(long, bool, double, std::optional<c10::ScalarType>, std::optional<c10::Layout>, std::optional<c10::Device>, std::optional<bool>) + 0x1d70 (0x7f28b0ae4e20 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #14: at::native::kaiser_window(long, std::optional<c10::ScalarType>, std::optional<c10::Layout>, std::optional<c10::Device>, std::optional<bool>) + 0x7a7 (0x7f28b0b41a67 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #15: <unknown function> + 0xa3fdf41 (0x7f28b5f0cf41 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #16: <unknown function> + 0xa3ff76f (0x7f28b5f0e76f in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #17: <unknown function> + 0x6668d51 (0x7f28b2177d51 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #18: <unknown function> + 0x666816d (0x7f28b217716d in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #19: at::_ops::kaiser_window::redispatch(c10::DispatchKeySet, long, std::optional<c10::ScalarType>, std::optional<c10::Layout>, std::optional<c10::Device>, std::optional<bool>) + 0x12da (0x7f28b4b3768a in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #20: <unknown function> + 0x94a385c (0x7f28b4fb285c in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #21: <unknown function> + 0x94a508f (0x7f28b4fb408f in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #22: <unknown function> + 0x6668d51 (0x7f28b2177d51 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #23: at::_ops::kaiser_window::call(long, std::optional<c10::ScalarType>, std::optional<c10::Layout>, std::optional<c10::Device>, std::optional<bool>) + 0x2b12 (0x7f28b485c0d2 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #24: ./reproduce_kaiser_window() [0x4063de]
frame #25: main + 0x110f (0x4034ef in ./reproduce_kaiser_window)
frame #26: <unknown function> + 0x29d90 (0x7f28a1d65d90 in /lib/x86_64-linux-gnu/libc.so.6)
frame #27: __libc_start_main + 0x80 (0x7f28a1d65e40 in /lib/x86_64-linux-gnu/libc.so.6)
frame #28: _start + 0x25 (0x402315 in ./reproduce_kaiser_window)

Aborted (core dumped)
```

### Versions

PyTorch version: 2.3.0a0+git03969cb
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 12.0.1
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 17:29:18) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-89-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             96
On-line CPU(s) list:                0-95
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          2
Stepping:                           7
CPU max MHz:                        4000.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           6000.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           48 MiB (48 instances)
L3 cache:                           71.5 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-23,48-71
NUMA node1 CPU(s):                  24-47,72-95
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] optree==0.10.0
[pip3] torch==2.3.0a0+git03969cb
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-include               2023.1.0         h06a4308_46344  
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] torch                     2.3.0a0+git03969cb           dev_0    <develop>

### PR DESCRIPTION
**Pull request by CaoE**
Fixes #117230.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 1]
```diff
@@ -523,8 +523,8 @@ static void kaiser_window_kernel(TensorIteratorBase& iter, int64_t window_length
     using opmath_t = at::opmath_type<scalar_t>;
     const opmath_t alpha = static_cast<opmath_t>((window_length - 1) / 2.0);
     const opmath_t beta_ = static_cast<opmath_t>(beta);
-    cpu_kernel(iter, [=](scalar_t a){
-        return calc_i0(beta_ * std::sqrt(1 - std::pow((static_cast<opmath_t>(a) - alpha) / alpha, static_cast<opmath_t>(2.0)))) / calc_i0(beta_);
+    cpu_kernel(iter, [=](scalar_t a) -> scalar_t {
+        return calc_i0(beta_ * std::sqrt(std::abs(1 - std::pow((static_cast<opmath_t>(a) - alpha) / alpha, static_cast<opmath_t>(2.0))))) / calc_i0(beta_);
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:527]: Is this `std::abs` necessary? `a` is non-negative here and `std::pow((static_cast<opmath_t>(a) - alpha) / alpha` is always >=0, right?

Comment (CaoE) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:527]: When data type is bf16, `1 - std::pow((static_cast<opmath_t>(a) - alpha) / alpha, static_cast<opmath_t>(2.0)))` may be < 0 as bf16 will lose accuracy.

Comment (jgong5) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:527]: But this formula `1 - std::pow((static_cast<opmath_t>(a) - alpha) / alpha, static_cast<opmath_t>(2.0)))` is of float type when `scalar_t` is `bfloat16`, right?

Comment (CaoE) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:527]: Yes. But scalar_t a has already loss accuracy. 
It indeed may generate negative values when dtype is bf16. https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorFactories.cpp#L1642

Comment (jgong5) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:527]: Make sense. Thanks.

**Review by jgong5 — APPROVED** [review]

**Comment by CaoE** [comment]
@soumith Could you please review this PR ? Thank you.

**Review by soumith — APPROVED** [review]

**Pull request closed without merging**